### PR TITLE
gcc: update to 14.1.0

### PR DESCRIPTION
**Pull request by MehdiChinoune**

### CODE REVIEW

**Comment by ognevny** [comment]
looks like gccrs is far to be a working compiler :)
```
gccrs -g -O2 -c main.rs -o test.o -frust-incomplete-and-experimental-compiler-do-not-use
main.rs:2:15: error: Cannot find path 'std' in this scope [E0433]
    2 |     let buf = std::io::stdout().lock().unwrap();
      |
```

**Comment by lazka** [comment]
@lhmouse suggested this patch on IRC: https://github.com/lhmouse/MINGW-packages/blob/e291e2a5457873aaf968202fd42708766a21fe05/mingw-w64-gcc/0101-fix-thiscall-attribute.patch regarding this bugreport https://gcc.gnu.org/bugzilla/show_bug.cgi?id=114968

**Comment by MehdiChinoune** [comment]
Let's wait for the patch to be reviewed and accepted first.

**Comment by lhmouse** [comment]
I think this should not cause usability issues for non-i686 platforms. Upgrading GCC for MINGW64 and UCRT64 should be acceptable.

**Comment by lhmouse** [comment]
Jakub Jalinek has posted a new patch: https://gcc.gnu.org/bugzilla/attachment.cgi?id=58123
Please try building GCC for MINGW32 and using the new compiler to rebuild boost.

I got random zombie processes every time when building GCC on msys2-runtime 3.5:
```
$ ps -lfe
lh_mouse    3014       1 cons0    11:15:42 /mingw32/i686-w64-mingw32/bin/as <defunct>
```
I have downgraded to msys2-runtime-3.4. Proceeding now.

**Comment by lhmouse** [comment]
All bootstrapped successfully with this patch: https://gcc.gnu.org/bugzilla/attachment.cgi?id=58125

MINGW32 GCC causes random segfaults in the assembler when building Rust, but I think it's unrelated to this patch.

**Comment by ognevny** [comment]
> MINGW32 GCC causes random segfaults in the assembler when building Rust, but I think it's unrelated to this patch.

huge warning: Rust thing on GCC is unstable and marked as not usable by the maintainers :)

**Comment by lhmouse** [comment]
> > MINGW32 GCC causes random segfaults in the assembler when building Rust, but I think it's unrelated to this patch.
> 
> huge warning: Rust thing on GCC is unstable and marked as not usable by the maintainers :)

Is it really acceptable to be just 'unstable' if it is causing segfaults in the assembler? 😂 
There is 1/3 of probability that it succeeds.

**Comment by ognevny** [comment]
> Is it really acceptable to be just 'unstable' if it is causing segfaults in the assembler? 😂 

maybe

**Comment by mati865** [comment]
> > > MINGW32 GCC causes random segfaults in the assembler when building Rust, but I think it's unrelated to this patch.
> > 
> > 
> > huge warning: Rust thing on GCC is unstable and marked as not usable by the maintainers :)
> 
> Is it really acceptable to be just 'unstable' if it is causing segfaults in the assembler? 😂 There is 1/3 of probability that it succeeds.

No matter what, assembler should not segfault.
So that could be a bug in Binutils or system/hardware issue.

**Comment by MehdiChinoune** [comment]
Rust is not enabled on mingw32.

**Comment by iahung3** [comment]
Add MINGWARM64 environment?

https://github.com/msys2/MINGW-packages/issues/20838

**Comment by lhmouse** [comment]
@MehdiChinoune Do you think the random segfaults in the MINGW32 assembler are related to https://github.com/msys2/MINGW-packages/commit/37ebf8cf8236d4dfbd6c7ec67f5003c505bde36d#diff-cf683a120abf18dd9f774ea5df77137d01e19f359f7cb40c1fc152355cde7b80L290 ?

**Comment by MehdiChinoune** [comment]
I don't know!

**Comment by lhmouse** [comment]
> I don't know!

Okay, making some random attempts to address the issue... I don't think there is anything wrong within GCC or GAS; it looks more likely a technical limitation as it only happens on MINGW32. I am trying `bcdedit /set IncreaseUserVa 3072` now.

**Comment by MehdiChinoune** [comment]
@lhmouse rust is not enabled on mingw32, so I think this PR is ready.
What do you think?

**Comment by lhmouse** [comment]
Looks good to me.

**Comment by MehdiChinoune** [comment]
> '2001-fix-building-rust-on-mingw-w64.patch' seems no longer necessary.

Are you talking about gcc-14 or gcc-15?

**Comment by lhmouse** [comment]
> > '2001-fix-building-rust-on-mingw-w64.patch' seems no longer necessary.
> 
> 
> 
> Are you talking about gcc-14 or gcc-15?

GCC 14 of course.

**Comment by lhmouse** [comment]
The segfault does not seem to have anything to do with Rust. With Rust disabled, I get similar errors in somewhere else:
```
/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/./prev-gcc/xg++ -B/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/./prev-gcc/ -B/mingw32/i686-w64-mingw32/bin/ -nostdinc++ -B/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/src/.libs -B/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/libsupc++/.libs  -I/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/include/i686-w64-mingw32  -I/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/include  -I/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/gcc/libstdc++-v3/libsupc++ -L/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/src/.libs -L/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/libsupc++/.libs   -g -O2 -D__USE_MINGW_ACCESS -Wno-pedantic-ms-format -DIN_GCC    -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Wconditionally-supported -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H  -DGENERATOR_FILE -static-libstdc++ -Wl,--stack,12582912 -Wl,--undefined=HOST_EXTRA_OBJS_SYMBOL -o build/genconditions.exe \
    build/genconditions.o build/rtl.o build/read-rtl.o build/ggc-none.o build/vec.o build/min-insn-modes.o build/gensupport.o build/print-rtl.o build/hash-table.o build/sort.o build/read-md.o build/errors.o .././libiberty/libiberty.a
make[3]: *** [Makefile:3196: build/genconditions.exe] Error 1
make[3]: *** Waiting for unfinished jobs....
collect2.exe: error: ld returned 5 exit status
/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/./prev-gcc/xg++ -B/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/./prev-gcc/ -B/mingw32/i686-w64-mingw32/bin/ -nostdinc++ -B/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/src/.libs -B/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/libsupc++/.libs  -I/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/include/i686-w64-mingw32  -I/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/include  -I/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/gcc/libstdc++-v3/libsupc++ -L/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/src/.libs -L/home/lh_mouse/GitHub/MINGW-packages/mingw-w64-gcc/src/build-MINGW32/prev-i686-w64-mingw32/libstdc++-v3/libsupc++/.libs -c   -g -O2 -D__USE_MINGW_ACCESS -Wno-pedantic-ms-format -DIN_GCC    -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Wconditionally-supported -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H  -DGENERATOR_FILE -I. -Ibuild -I../../gcc/gcc -I../../gcc/gcc/build -I../../gcc/gcc/../include  -I../../gcc/gcc/../libcpp/include -D__USE_MINGW_ANSI_STDIO=1 -DCOM_NO_WINDOWS_H -I/mingw32/include \
        -o build/genextract.o ../../gcc/gcc/genextract.cc
make[3]: *** [Makefile:3196: build/genextract.exe] Error 1
```

Running these commands by hand does not fail. These errors do not happen on MINGW64 or UCRT64.

File: mingw-w64-gcc/PKGBUILD [hunk 1]
```diff
@@ -5,10 +5,17 @@
...
-_enable_rust=no
-_enable_jit=yes
+if [[ ${CARCH} != i686 ]]; then
+  _enable_ada=yes
+  _enable_objc=yes
+  _enable_rust=yes
+  _enable_jit=yes
+else
+  _enable_ada=no
+  _enable_objc=no
```
Comment (hmelder) [inline mingw-w64-gcc/PKGBUILD:15]: What's the reason for disabling gcc-objc on i686?

Comment (hmelder) [reply mingw-w64-gcc/PKGBUILD:15]: This broke our CI pipeline for GNUstep Foundation

Comment (MehdiChinoune) [reply mingw-w64-gcc/PKGBUILD:15]: We started slowly dropping 32-bit packages.
We updated gcc to 14.1.0 moths ago.

The solution to your issue is to drop i686.

**Pull request merged**